[multistage][bug] multistage engine leaf stage returns incompatible data type.

Problematic GROUP BY V1 behavior returns STRING instead of INT as indicated in schema
see: https://github.com/apache/pinot/actions/runs/3688265209/jobs/6242862869

This is most likely due to group by results when empty doesn't know exactly what the return type of group by key column is, thus created STRING placeholder return instead. 

```
2022-12-13T18:43:31.7985289Z [ERROR] org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testGeneratedQueries  Time elapsed: 14.5 s  <<< FAILURE!
2022-12-13T18:43:31.7985807Z java.lang.AssertionError: 
2022-12-13T18:43:31.7986071Z Caught exception while testing query!
2022-12-13T18:43:31.7986778Z Pinot query: SELECT DepartureDelayGroups, Flights FROM mytable WHERE FirstDepTime > 1504 AND DepDel15 BETWEEN 0 AND -9999 AND DestWac BETWEEN 21 AND 51 GROUP BY DepartureDelayGroups, Flights LIMIT 10000
2022-12-13T18:43:31.7987617Z H2 query: SELECT DepartureDelayGroups, Flights FROM mytable WHERE FirstDepTime > 1504 AND DepDel15 BETWEEN 0 AND -9999 AND DestWac BETWEEN 21 AND 51 GROUP BY DepartureDelayGroups, Flights LIMIT 10000
2022-12-13T18:43:31.7988094Z 	at org.testng.Assert.fail(Assert.java:82)
2022-12-13T18:43:31.7988602Z 	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.failure(ClusterIntegrationTestUtils.java:889)
2022-12-13T18:43:31.7989290Z 	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:568)
2022-12-13T18:43:31.7990007Z 	at org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testQuery(MultiStageEngineIntegrationTest.java:128)
2022-12-13T18:43:31.7990760Z 	at org.apache.pinot.integration.tests.BaseClusterIntegrationTestSet.testGeneratedQueries(BaseClusterIntegrationTestSet.java:448)
2022-12-13T18:43:31.7991569Z 	at org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testGeneratedQueries(MultiStageEngineIntegrationTest.java:99)

...

2022-12-13T18:43:31.8009092Z Caused by: java.lang.RuntimeException: Got Exceptions from Query Response: {"exceptions":[{"errorCode":200,"message":"QueryExecutionError:\njava.lang.RuntimeException: Received error query execution result block: {1000=Incompatible selection result data schema:  Expected: [DepartureDelayGroups(INT),Flights(INT)]. Actual: [DepartureDelayGroups(STRING),Flights(STRING)]\njava.lang.IllegalStateException: Incompatible selection result data schema:  Expected: [DepartureDelayGroups(INT),Flights(INT)]. Actual: [DepartureDelayGroups(STRING),Flights(STRING)]\n\tat com.google.common.base.Preconditions.checkState(Preconditions.java:444)\n\tat org.apache.pinot.query.runtime.operator.LeafStageTransferableBlockOperator.composeDistinctTransferableBlock(LeafStageTransferableBlockOperator.java:150)\n\tat org.apache.pinot.query.runtime.operator.LeafStageTransferableBlockOperator.composeTransferableBlock(LeafStageTransferableBlockOperator.java:129)"}],"numServersQueried":0,"numServersResponded":0,"numSegmentsQueried":0,"numSegmentsProcessed":0,"numSegmentsMatched":0,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"numDocsScanned":0,"numEntriesScannedInFilter":0,"numEntriesScannedPostFilter":0,"numGroupsLimitReached":false,"totalDocs":0,"timeUsedMs":0,"offlineThreadCpuTimeNs":0,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":0,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":0,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":0,"realtimeTotalCpuTimeNs":0,"segmentStatistics":[],"traceInfo":{},"minConsumingFreshnessTimeMs":0,"numRowsResultSet":0,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0}
2022-12-13T18:43:31.8012893Z 	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQueryInternal(ClusterIntegrationTestUtils.java:579)
2022-12-13T18:43:31.8013695Z 	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:566)
2022-12-13T18:43:31.8014265Z 	... 33 more
2022-12-13T18:43:31.8014944Z 
2022-12-13T18:43:31.8015719Z [ERROR] org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testHardcodedQueriesMultiStage  Time elapsed: 4.356 s
2022-12-13T18:43:32.9604692Z [INFO] Running org.apache.pinot.integration.tests.ControllerPeriodicTasksIntegrationTest
2022-12-13T18:44:59.5704240Z Dec 13, 2022 6:43:46 PM org.glassfish.grizzly.http.server.NetworkListener start\
```

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/AggregationResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/AggregationResultsBlock.java
@@ -57,6 +57,11 @@ public class AggregationResultsBlock extends BaseResultsBlock {
   }
 
   @Override
+  public long getNumRecords() {
+    return 1;
+  }
+
+  @Override
   public DataSchema getDataSchema(QueryContext queryContext) {
     boolean returnFinalResult = queryContext.isServerReturnFinalResult();
     int numColumns = _aggregationFunctions.length;

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/AggregationResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/AggregationResultsBlock.java
@@ -57,7 +57,7 @@ public class AggregationResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return 1;
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/BaseResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/BaseResultsBlock.java
@@ -159,6 +159,11 @@ public abstract class BaseResultsBlock implements Block {
   }
 
   /**
+   * Indicate the size of the result.
+   */
+  public abstract long getNumRecords();
+
+  /**
    * Returns the data schema for the results. Return {@code null} when the block only contains metadata.
    */
   @Nullable

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/BaseResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/BaseResultsBlock.java
@@ -159,9 +159,11 @@ public abstract class BaseResultsBlock implements Block {
   }
 
   /**
-   * Indicate the size of the result.
+   * Returns the total size (number of rows) in this result block, without having to materialize the rows.
+   *
+   * @see BaseResultsBlock#getRows(QueryContext)
    */
-  public abstract long getNumRecords();
+  public abstract int getNumRows();
 
   /**
    * Returns the data schema for the results. Return {@code null} when the block only contains metadata.

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/DistinctResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/DistinctResultsBlock.java
@@ -52,6 +52,11 @@ public class DistinctResultsBlock extends BaseResultsBlock {
   }
 
   @Override
+  public long getNumRecords() {
+    return _distinctTable.size();
+  }
+
+  @Override
   public DataSchema getDataSchema(QueryContext queryContext) {
     return _distinctTable.getDataSchema();
   }

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/DistinctResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/DistinctResultsBlock.java
@@ -52,7 +52,7 @@ public class DistinctResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return _distinctTable.size();
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExceptionResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExceptionResultsBlock.java
@@ -38,6 +38,11 @@ public class ExceptionResultsBlock extends BaseResultsBlock {
     this(QueryException.QUERY_EXECUTION_ERROR, t);
   }
 
+  @Override
+  public long getNumRecords() {
+    return 0;
+  }
+
   @Nullable
   @Override
   public DataSchema getDataSchema(QueryContext queryContext) {

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExceptionResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExceptionResultsBlock.java
@@ -39,7 +39,7 @@ public class ExceptionResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return 0;
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExplainResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExplainResultsBlock.java
@@ -47,7 +47,7 @@ public class ExplainResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return _entries.size();
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExplainResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/ExplainResultsBlock.java
@@ -47,6 +47,11 @@ public class ExplainResultsBlock extends BaseResultsBlock {
   }
 
   @Override
+  public long getNumRecords() {
+    return _entries.size();
+  }
+
+  @Override
   public Collection<Object[]> getRows(QueryContext queryContext) {
     List<Object[]> rows = new ArrayList<>(_entries.size());
     for (ExplainEntry entry : _entries) {

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/GroupByResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/GroupByResultsBlock.java
@@ -136,6 +136,11 @@ public class GroupByResultsBlock extends BaseResultsBlock {
     _resizeTimeMs = resizeTimeMs;
   }
 
+  @Override
+  public long getNumRecords() {
+    return _table == null ? 0 : _table.size();
+  }
+
   @Nullable
   @Override
   public DataSchema getDataSchema(QueryContext queryContext) {

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/GroupByResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/GroupByResultsBlock.java
@@ -137,7 +137,7 @@ public class GroupByResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return _table == null ? 0 : _table.size();
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/MetadataResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/MetadataResultsBlock.java
@@ -29,7 +29,7 @@ import org.apache.pinot.core.query.request.context.QueryContext;
 public class MetadataResultsBlock extends BaseResultsBlock {
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return 0;
   }
 

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/MetadataResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/MetadataResultsBlock.java
@@ -28,6 +28,11 @@ import org.apache.pinot.core.query.request.context.QueryContext;
 
 public class MetadataResultsBlock extends BaseResultsBlock {
 
+  @Override
+  public long getNumRecords() {
+    return 0;
+  }
+
   @Nullable
   @Override
   public DataSchema getDataSchema(QueryContext queryContext) {

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/SelectionResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/SelectionResultsBlock.java
@@ -63,6 +63,11 @@ public class SelectionResultsBlock extends BaseResultsBlock {
   }
 
   @Override
+  public long getNumRecords() {
+    return _rows.size();
+  }
+
+  @Override
   public DataSchema getDataSchema(QueryContext queryContext) {
     return _dataSchema;
   }

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/SelectionResultsBlock.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/results/SelectionResultsBlock.java
@@ -63,7 +63,7 @@ public class SelectionResultsBlock extends BaseResultsBlock {
   }
 
   @Override
-  public long getNumRecords() {
+  public int getNumRows() {
     return _rows.size();
   }
 

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperator.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperator.java
@@ -93,7 +93,7 @@ public class LeafStageTransferableBlockOperator extends BaseOperator<Transferabl
     } else {
       if (_currentIndex < _baseResultBlock.size()) {
         InstanceResponseBlock responseBlock = _baseResultBlock.get(_currentIndex++);
-        if (responseBlock.getResultsBlock() != null && responseBlock.getResultsBlock().getNumRecords() > 0) {
+        if (responseBlock.getResultsBlock() != null && responseBlock.getResultsBlock().getNumRows() > 0) {
           return composeTransferableBlock(responseBlock, _desiredDataSchema);
         } else {
           return new TransferableBlock(Collections.emptyList(), _desiredDataSchema, DataBlock.Type.ROW);

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperator.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperator.java
@@ -93,7 +93,7 @@ public class LeafStageTransferableBlockOperator extends BaseOperator<Transferabl
     } else {
       if (_currentIndex < _baseResultBlock.size()) {
         InstanceResponseBlock responseBlock = _baseResultBlock.get(_currentIndex++);
-        if (responseBlock.getResultsBlock() != null) {
+        if (responseBlock.getResultsBlock() != null && responseBlock.getResultsBlock().getNumRecords() > 0) {
           return composeTransferableBlock(responseBlock, _desiredDataSchema);
         } else {
           return new TransferableBlock(Collections.emptyList(), _desiredDataSchema, DataBlock.Type.ROW);

--- a/pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperatorTest.java
+++ b/pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperatorTest.java
@@ -262,4 +262,70 @@ public class LeafStageTransferableBlockOperatorTest {
     // Then:
     Assert.assertFalse(resultBlock.isErrorBlock());
   }
+
+  @Test
+  public void shouldNotErrorOutWhenIncorrectDataSchemaProvidedWithEmptyRowsSelection() {
+    // Given:
+    QueryContext queryContext = QueryContextConverterUtils.getQueryContext("SELECT strCol, intCol FROM tbl");
+    DataSchema resultSchema = new DataSchema(new String[]{"strCol", "intCol"},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.STRING});
+    DataSchema desiredSchema = new DataSchema(new String[]{"strCol", "intCol"},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.INT});
+
+    // When:
+    List<InstanceResponseBlock> responseBlockList = Collections.singletonList(
+        new InstanceResponseBlock(new SelectionResultsBlock(resultSchema, Collections.emptyList()), queryContext));
+    LeafStageTransferableBlockOperator operator = new LeafStageTransferableBlockOperator(responseBlockList,
+        desiredSchema);
+    TransferableBlock resultBlock = operator.nextBlock();
+
+    // Then:
+    Assert.assertEquals(resultBlock.getContainer().size(), 0);
+    Assert.assertEquals(resultBlock.getDataSchema(), desiredSchema);
+  }
+
+  @Test
+  public void shouldNotErrorOutWhenIncorrectDataSchemaProvidedWithEmptyRowsDistinct() {
+    // Given:
+    QueryContext queryContext = QueryContextConverterUtils.getQueryContext(
+        "SELECT strCol, intCol FROM tbl GROUP BY strCol, intCol");
+    DataSchema resultSchema = new DataSchema(new String[]{"strCol", "intCol"},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.STRING});
+    DataSchema desiredSchema = new DataSchema(new String[]{"strCol", "intCol"},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.INT});
+
+    // When:
+    List<InstanceResponseBlock> responseBlockList = Collections.singletonList(
+        new InstanceResponseBlock(new DistinctResultsBlock(mock(DistinctAggregationFunction.class),
+            new DistinctTable(resultSchema, Collections.emptyList())), queryContext));
+    LeafStageTransferableBlockOperator operator = new LeafStageTransferableBlockOperator(responseBlockList,
+        desiredSchema);
+    TransferableBlock resultBlock = operator.nextBlock();
+
+    // Then:
+    Assert.assertEquals(resultBlock.getContainer().size(), 0);
+    Assert.assertEquals(resultBlock.getDataSchema(), desiredSchema);
+  }
+
+  @Test
+  public void shouldNotErrorOutWhenIncorrectDataSchemaProvidedWithEmptyRowsGroupBy() {
+    // Given:
+    QueryContext queryContext = QueryContextConverterUtils.getQueryContext(
+        "SELECT strCol, SUM(intCol) FROM tbl GROUP BY strCol");
+    DataSchema resultSchema = new DataSchema(new String[]{"strCol", "SUM(intCol)"},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.STRING});
+    DataSchema desiredSchema = new DataSchema(new String[]{"strCol", "SUM(intCol)"},
+      new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.STRING, DataSchema.ColumnDataType.INT});
+
+    // When:
+    List<InstanceResponseBlock> responseBlockList = Collections.singletonList(
+        new InstanceResponseBlock(new GroupByResultsBlock(resultSchema, Collections.emptyList()), queryContext));
+    LeafStageTransferableBlockOperator operator = new LeafStageTransferableBlockOperator(responseBlockList,
+        desiredSchema);
+    TransferableBlock resultBlock = operator.nextBlock();
+
+    // Then:
+    Assert.assertEquals(resultBlock.getContainer().size(), 0);
+    Assert.assertEquals(resultBlock.getDataSchema(), desiredSchema);
+  }
 }

